gamemodes/terrortown/gamemode/server/sv_player_ext.lua

```diff
@@ -1311,6 +1311,132 @@ function plymeta:SafeDropWeapon(wep, keepSelection)
 	return true
 end
 
+local function TraceAmmoDrop(ply)
+	local pos, ang = ply:GetShootPos(), ply:EyeAngles()
+	local fwd, rgt, up = ang:Forward(), ang:Right(), ang:Up()
+
+	local dir = fwd * 32
+	rgt:Mul(6)
+	up:Mul(-5)
+	dir:Add(rgt)
+	dir:Add(up)
+
+	local tr = util.QuickTrace(pos, dir, ply)
+
+	return tr, pos, dir, fwd
+end
+
+---
+-- Checks if the ammo can be dropped in a safe manner.
+-- @param Weapon wep The weapon that should be dropped
+-- @return boolean Returns if this weapon can be dropped
+-- @realm server
+function plymeta:CanSafeDropAmmo(wep)
+	if not IsValid(self) or not (IsValid(wep) and wep.AmmoEnt) then
+		return false
+	end
+
+	local tr = TraceAmmoDrop(self)
+
+	if tr.HitWorld then
+		LANG.Msg(self, "drop_no_room_ammo", nil, MSG_MSTACK_WARN)
+
+		return false
+	end
+
+	return true
+end
+
+---
+-- Called to drop ammo in a safe manner (e.g. preparing and space-check).
+-- @param Weapon wep The weapon that should be referenced when dropping ammo.
+-- @param boolean useClip If set to true the ammo is dropped from the clip, otherwise, from reserve ammo.
+-- @return boolean Returns if ammo is dropped
+-- @realm server
+function plymeta:SafeDropAmmo(wep, useClip)
+	if not self:CanSafeDropAmmo(wep) then
+		return false
+	end
+
+	-- most of this goes into computing ammo amount
+	local amt = 0
+	if useClip then
+		amt = wep:Clip1()
+	else
+		amt = math.min(
+			wep.Primary.ClipSize,
+			self:GetAmmoCount(wep.Primary.Ammo)
+		)
+	end
+	local hook_data = {amt}
+
+	---
+	-- @realm server
+	if hook.Run("TTT2DropAmmo", self, hook_data) == false then
+		LANG.Msg(self, useClip and "drop_ammo_prevented" or "drop_reserve_prevented", nil, MSG_MSTACK_WARN)
+
+		return false
+	end
+	amt = hook_data[1]
+	if amt < 1 or amt <= wep.Primary.ClipSize * 0.25 then
+		LANG.Msg(self, useClip and "drop_no_ammo" or "drop_no_reserve", nil, MSG_MSTACK_WARN)
+
+		return false
+	end
+
+	self:AnimPerformGesture(ACT_GMOD_GESTURE_ITEM_GIVE)
+
+	WEPS.DropNotifiedAmmo(self, wep, amt)
+
+	if useClip then
+		wep:SetClip1( math.max(wep:Clip1() - amt, 0) )
+	else
+		self:RemoveAmmo(amt, wep.Primary.Ammo)
+	end
+
+	return true
+end
+
+---
+-- Called to drop a weapon in a safe manner (e.g. preparing and space-check).
+-- @param Weapon wep The weapon that should be dropped
+-- @param boolean keepSelection If set to true the current selection is kept if not dropped
+-- @return boolean Returns if this weapon is dropped
+-- @realm server
+function plymeta:DropAmmo(wep, amt)
```

Comment:
Why do we want this function to be exposed to developers? 
Shouldn't we make sure that they use the safe drop all the time?

Im bringing this up, because this function is missing some stuff that the safedrop function currently has eg. The animation etc. if we would just differentiate between safe and non safe drops, i would expect the normal drop function to take care of most of the logic except the safe check call.

Reply:
Either I'm missing something, or you're talking about `plymeta:SafeDrioWeapon`, but this is about ammo

Reply:
I wrote it this way because it is meant to be a parallel to [Player:DropWeapon](https://wiki.facepunch.com/gmod/Player:DropWeapon), as a majority of this code exists to mirror the architecture of similar functions for dropping weapons. The reason this one stands out, is that it has no primitive equivalent.

Reply:
I don't see a reason to keep this extra function and would rather merge both together. Because we would not want a user to actually use the DropAmmo function, do we?

Reply:
I think the whole point of this PR is to add an option to drop ammo in a reasonable way. It was possible before, but it took the ammo out of the weapon instead of your storage afaik

Reply:
I understand the PR and the reason im only questioning the existence of two functions where the latter is rather incomplete and does not make sense to call on its own

Reply:
My only personal justification is that sometimes addons have called DropWeapon, when SafeDropWeapon would have stopped them, and it was beneficial because the collision check is too tight in some circumstances (especially if you're crouching, looking at a weapon on the floor) -- or in situations where the grenade disarms should always be able to force drop a weapon rather than letting people keep it if they keep their face snug to a corner. Some sort of effect which forcibly spills a player's ammo regardless of the room to do it wouldn't be unreasonable to imagine.

Reply:
Thats why I asked. If you only want to do that, then move the calculation and all other logic from the safe function to the drop function and only do the safety check in the safedrop function. In this way their responsibilities are much more clear and are closer to what someone would think they do.

Reply:
which line ranges in particular do you think should be moved? i'll need to change some signatures potentially

Reply:
Everything except for the candrop check

Reply:
I think that did it? The only thing now is that ply:DropAmmo could still be interrupted, and show MSTACK messages even if it doesn't originate as a user-driven intention.